enterprise/cypher/compiled-expressions/src/main/scala/org/neo4j/cypher/internal/runtime/compiled/expressions/IntermediateCodeGeneration.scala

```diff
@@ -0,0 +1,235 @@
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("cos"), arg))
+          case _ => None
+        }
+      case functions.Cot =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("cot"), arg))
+          case _ => None
+        }
+      case functions.Asin =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("asin"), arg))
+          case _ => None
+        }
+      case functions.Haversin =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("haversin"), arg))
+          case _ => None
+        }
+      case functions.Sin =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("sin"), arg))
+          case _ => None
+        }
+      case functions.Atan =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("atan"), arg))
+          case _ => None
+        }
+      case functions.Atan2 =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("atan2"), arg))
+          case _ => None
+        }
+      case functions.Tan =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("tan"), arg))
+          case _ => None
+        }
+      case functions.Round =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("round"), arg))
+          case _ => None
+        }
+      case functions.Rand =>
+        Some(invokeStatic(method[CypherFunctions, DoubleValue]("rand")))
+      case functions.Abs =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("abs"), arg))
+          case _ => None
+        }
+      case functions.Ceil =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("ceil"), arg))
+          case _ => None
+        }
+      case functions.Floor =>
+        compile(c.args.head) match {
+          case Some(arg) =>
+            Some(invokeStatic(method[CypherFunctions, Value, AnyValue]("floor"), arg))
+          case _ => None
+        }
+
+      case _ => None
+    }
+
+    //math
+    case Multiply(lhs, rhs) =>
+      (compile(lhs), compile(rhs)) match {
+        case (Some(l), Some(r)) =>
+          Some(invokeStatic(method[CypherMath, AnyValue, AnyValue, AnyValue]("multiply"), l, r))
+
+        case _ => None
+      }
```

Comment:
And for the binary functions we could use the `for-yield` construct. It's pretty awesome for all kinds of `Option` and `Seq` manipulations, but I haven't seen it much in the codebase. Goes something like 

```
case Multiply(lhs, rhs) =>
  for { l <- compile(lhs)
        r <- compile(rhs) 
  } yield invokeStatic(method[CypherMath, AnyValue, AnyValue, AnyValue]("multiply"), l, r)
```

and gives the same outcome. Note how the expression after yield automatically gets wrapped in an option, but only if both `compile(lhs)` and `compile(rhs)` are non-empty. Also `compile(rhs)` will not get called if `compile(lhs)` is non-empty.

Reply:
yes this is nicer, I have updated the code.